// feedr/out/lib/plugins.js
// Generated by CoffeeScript 1.9.1
(function() {
  module.exports = {
    github: {
      check: function(arg, next) {
        var data, failedResponseError, feed;
        feed = arg.feed, data = arg.data;
        if (feed.url.indexOf('github.com') !== -1 && (data != null ? data.message : void 0)) {
          failedResponseError = new Error(data.message);
          return next(failedResponseError);
        }
        return next();
      }
    },
    xml: {
      parse: function(arg, next) {
        var data, err, feed, feedr, parser, ref, response, xml2js, xml2jsOptions;
        feedr = arg.feedr, feed = arg.feed, response = arg.response, data = arg.data;
        if (!(feed.parse === 'xml' || ((ref = feed.extension) === '.xml' || ref === '.atom' || ref === '.rss' || ref === '.rdf' || ref === '.html' || ref === '.html') || response.headers['content-type'].indexOf('xml') !== -1 || response.headers['content-type'].indexOf('html') !== -1)) {
          return next();
        }
        xml2jsOptions = require('extendr').deepExtend({}, feedr.config.xml2jsOptions || {}, feed.xml2jsOptions || {});
        xml2js = require('xml2js');
        parser = new xml2js.Parser(xml2jsOptions);
        parser.on('end', function(data) {
          return next(null, data);
        });
        try {
          return parser.parseString(data.toString().trim());
        } catch (_error) {
          err = _error;
          return next(err);
        }
      }
    },
    cson: {
      parse: function(arg, next) {
        var data, feed, ref, response;
        feed = arg.feed, response = arg.response, data = arg.data;
        if (!(feed.parse === 'cson' || ((ref = feed.extension) === '.coffee' || ref === '.cson') || response.headers['content-type'].indexOf('coffeescript') !== -1 || response.headers['content-type'].indexOf('cson') !== -1)) {
          return next();
        }
        return require('CSON').parseCSONString(data.toString(), next);
      }
    },
    json: {
      parse: function(arg, next) {
        var data, err, feed, feedr, ref, response;
        feedr = arg.feedr, feed = arg.feed, response = arg.response, data = arg.data;
        if (!(feed.parse === 'json' || ((ref = feed.extension) === '.json' || ref === '.jsonp' || ref === '.js') || response.headers['content-type'].indexOf('javascript') !== -1 || response.headers['content-type'].indexOf('json') !== -1)) {
          return next();
        }
        data = data.toString().trim().replace(/(^([\s\;]|\/\*\*\/)+|[\s\;]+$)/g, '');
        data = data.replace(/^[a-z0-9]+/gi, '').replace(/^\(|\)$/g, '');
        try {
          data = JSON.parse(data);
        } catch (_error) {
          err = _error;
          data = data.replace(/\\'/g, "'");
          try {
            data = JSON.parse(data);
          } catch (_error) {
            err = _error;
            return next(err);
          }
        }
        if (feed.clean) {
          feedr.log('debug', "Feedr is cleaning data from [" + feed.url + "]");
          data = feedr.cleanData(data);
        }
        return next(null, data);
      }
    },
    yaml: {
      parse: function(arg, next) {
        var data, err, feed, ref, response;
        feed = arg.feed, response = arg.response, data = arg.data;
        if (!(feed.parse === 'yaml' || ((ref = feed.extension) === '.yml' || ref === '.yaml') || response.headers['content-type'].indexOf('yaml') !== -1)) {
          return next();
        }
        try {
          data = require('js-yaml').load(data.toString().trim());
        } catch (_error) {
          err = _error;
          return next(err);
        }
        return next(null, data);
      }
    },
    string: {
      parse: function(arg, next) {
        var data, feed;
        feed = arg.feed, data = arg.data;
        if (!require('istextorbinary').isTextSync(feed.basename, data)) {
          return next();
        }
        return next(null, data.toString());
      }
    }
  };

}).call(this);
