// projectz/out/lib/backer-util.js
// Generated by CoffeeScript 1.9.1
(function() {
  var backerUtil, badgeUtil, projectzUtil;

  projectzUtil = require('./projectz-util');

  badgeUtil = require('./badge-util');

  module.exports = backerUtil = {
    getSponsorsText: function(opts) {
      var result, sponsor;
      result = '';
      if (opts.sponsors.length === 0) {
        result += "No sponsors yet! Will you be the first?\n\n";
        result += badgeUtil.getTypedBadges('donation', opts);
      } else {
        result += "These amazing people have contributed finances to this project:\n\n";
        result += '- ' + ((function() {
          var i, len, ref, results;
          ref = opts.sponsors;
          results = [];
          for (i = 0, len = ref.length; i < len; i++) {
            sponsor = ref[i];
            results.push(projectzUtil.getPersonText(sponsor));
          }
          return results;
        })()).join('\n- ');
        result += "\n\nBecome a sponsor!\n\n" + (badgeUtil.getTypedBadges('donation', opts));
      }
      return result;
    },
    getMaintainersText: function(opts) {
      var maintainer, result;
      result = '';
      if (opts.maintainers.length === 0) {
        result += "No maintainers yet! Will you be the first?";
      } else {
        result += "These amazing people are maintaining this project:\n\n";
        result += '- ' + ((function() {
          var i, len, ref, results;
          ref = opts.maintainers;
          results = [];
          for (i = 0, len = ref.length; i < len; i++) {
            maintainer = ref[i];
            results.push(projectzUtil.getPersonText(maintainer));
          }
          return results;
        })()).join('\n- ');
      }
      return result;
    },
    getContributorsText: function(opts) {
      var contributor, result;
      result = '';
      if (opts.contributors.length === 0) {
        result += "No contributors yet! Will you be the first?";
        result += "\n" + (backerUtil.getContributeText(opts));
      } else {
        result += "These amazing people have contributed code to this project:\n\n";
        result += '- ' + ((function() {
          var i, len, ref, results;
          ref = opts.contributors;
          results = [];
          for (i = 0, len = ref.length; i < len; i++) {
            contributor = ref[i];
            results.push(projectzUtil.getPersonText(contributor) + (" — [view contributions](https://github.com/" + opts.repo + "/commits?author=" + contributor.username + ")"));
          }
          return results;
        })()).join('\n- ');
        result += "\n\n" + (backerUtil.getContributeText(opts, 'Become a contributor!'));
      }
      return result;
    },
    getContributeText: function(opts, text) {
      var file, url;
      file = 'CONTRIBUTING.md';
      url = projectzUtil.getFileUrl(opts, file);
      if (text == null) {
        text = "Discover how you can contribute by heading on over to the `" + file + "` file.";
      }
      return "[" + text + "](" + url + ")";
    },
    getBackerSection: function(opts) {
      var result;
      if (opts == null) {
        opts = {};
      }
      if (!opts.licenses) {
        return '';
      }
      result = "## Backers\n\n### Maintainers\n\n" + (backerUtil.getMaintainersText(opts)) + "\n\n### Sponsors\n\n" + (backerUtil.getSponsorsText(opts)) + "\n\n### Contributors\n\n" + (backerUtil.getContributorsText(opts));
      return result;
    },
    getBackerFile: function(opts) {
      var result;
      if (opts == null) {
        opts = {};
      }
      if (!opts.licenses) {
        return '';
      }
      result = "# Backers\n\n## Maintainers\n\n" + (backerUtil.getMaintainersText(opts)) + "\n\n## Sponsors\n\n" + (backerUtil.getSponsorsText(opts)) + "\n\n## Contributors\n\n" + (backerUtil.getContributorsText(opts));
      return result;
    },
    getContributeSection: function(opts) {
      var result;
      result = "## Contribute\n\n" + (backerUtil.getContributeText(opts));
      return result;
    }
  };

}).call(this);
