// projectz/out/lib/license-util.js
// Generated by CoffeeScript 1.9.1
(function() {
  var categories, licenseUtil, licenses, projectzUtil;

  projectzUtil = require('./projectz-util');

  licenses = {
    'cc-by-4.0': {
      title: 'Creative Commons Attribution 4.0 International License',
      url: 'http://creativecommons.org/licenses/by/4.0/',
      body: 'See http://creativecommons.org/licenses/by/4.0/ for full legal text.'
    },
    'mit': {
      title: 'MIT License',
      category: 'permissive',
      url: 'http://opensource.org/licenses/mit-license.php',
      body: "Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the 'Software'), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:\n\nThe above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.\n\nTHE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE."
    }
  };

  categories = {
    permissive: {
      description: 'incredibly [permissive](http://en.wikipedia.org/wiki/Permissive_free_software_licence)'
    }
  };

  module.exports = licenseUtil = {
    getAuthorText: function(author) {
      if (author == null) {
        author = {};
      }
      return "Copyright &copy; " + (projectzUtil.getPersonText(author));
    },
    getAuthors: function(opts) {
      var author, entry, result;
      result = '';
      if (opts.authors.length !== 0) {
        result += '- ' + ((function() {
          var i, len, ref, results;
          ref = opts.authors;
          results = [];
          for (i = 0, len = ref.length; i < len; i++) {
            author = ref[i];
            results.push(entry = licenseUtil.getAuthorText(author));
          }
          return results;
        })()).join('\n- ');
      }
      return result;
    },
    getDescriptionsForLicenses: function(opts) {
      var category, categoryDescription, entry, license, result, scope, title, type, url;
      result = '';
      if (opts.licenses.length !== 0) {
        result += '- ' + ((function() {
          var i, len, ref, ref1, ref2, results;
          ref = opts.licenses;
          results = [];
          for (i = 0, len = ref.length; i < len; i++) {
            license = ref[i];
            entry = '';
            type = license.type, scope = license.scope;
            type = type.toLowerCase();
            if (licenses[type]) {
              entry += 'The';
              ref1 = licenses[type], title = ref1.title, url = ref1.url, category = ref1.category;
              categoryDescription = (ref2 = categories[category]) != null ? ref2.description : void 0;
              if (categoryDescription) {
                entry += ' ' + categoryDescription;
              }
              entry += ' [' + title + '](' + url + ')';
              if (scope) {
                entry += ' for ' + scope;
              }
            } else {
              entry += 'The #{type} License';
            }
            results.push(entry);
          }
          return results;
        })()).join('\n- ');
      }
      return result;
    },
    getBodiesForLicenses: function(opts) {
      var body, entry, license, result, scope, title, type;
      result = '';
      if (opts.licenses.length !== 0) {
        result += ((function() {
          var i, len, ref, ref1, results;
          ref = opts.licenses;
          results = [];
          for (i = 0, len = ref.length; i < len; i++) {
            license = ref[i];
            entry = '';
            type = license.type, scope = license.scope;
            type = type.toLowerCase();
            if (licenses[type]) {
              entry += 'The';
              ref1 = licenses[type], title = ref1.title, body = ref1.body;
              results.push(entry = '## ' + title + '\n' + body);
            } else {
              results.push(entry += '## The #{type} License\nMore information possibly available through the the [Open Source Initiative License Listing](http://opensource.org/licenses/alphabetical)');
            }
          }
          return results;
        })()).join('\n\n');
      }
      return result;
    },
    getLicenseIntroduction: function(opts) {
      var result;
      if (opts == null) {
        opts = {};
      }
      if (!opts.licenses) {
        return '';
      }
      result = "Unless stated otherwise all works are:\n\n" + (licenseUtil.getAuthors(opts)) + "\n\nand licensed under:\n\n" + (licenseUtil.getDescriptionsForLicenses(opts));
      return result;
    },
    getLicenseSection: function(opts) {
      var result;
      if (opts == null) {
        opts = {};
      }
      if (!opts.licenses) {
        return '';
      }
      result = "## License\n\n" + (licenseUtil.getLicenseIntroduction(opts));
      return result;
    },
    getLicenseFile: function(opts) {
      var result;
      if (opts == null) {
        opts = {};
      }
      if (!opts.licenses) {
        return '';
      }
      result = "# License\n\n" + (licenseUtil.getLicenseIntroduction(opts)) + "\n\n" + (licenseUtil.getBodiesForLicenses(opts));
      return result;
    }
  };

}).call(this);
